### PR TITLE
UI: DataTable, port from trunk; include URLBuilder, ViewControls, restructured Inputs

### PR DESCRIPTION
**Pull request by nhaagen**
This adds the DataTable and InputViewControls to v8.

### CODE REVIEW

**Comment by klees** [comment]
@nhaagen Do we understand why the tests don't run? @mjansenDatabay plz halp.

**Comment by mjansenDatabay** [comment]
> @nhaagen Do we understand why the tests don't run? @mjansenDatabay plz halp.

Here I am.

```
        _____________________________________________
      //:::::::::::::::::::::::::::::::::::::::::::::\\
    //:::_______:::::::::________::::::::::_____:::::::\\
  //:::_/   _-"":::_--"""        """--_::::\_  ):::::::::\\
 //:::/    /:::::_"                    "-_:::\/:::::|^\:::\\
//:::/   /~::::::I__                      \:::::::::|  \:::\\
\\:::\   (::::::::::""""---___________     "--------"  /::://
 \\:::\  |::::::::::::::::::::::::::::""""==____      /::://
  \\:::"\/::::::::::::::::::::::::::::::::::::::\   /~::://
    \\:::::::::::::::::::::::::::::::::::::::::::)/~::://
      \\::::\""""""------_____::::::::::::::::::::::://
        \\:::"\               """""-----_____:::::://
          \\:::"\    __----__                )::://
            \\:::"\/~::::::::~\_         __/~:://
              \\::::::::::::::::""----""":::://
                \\::::::::::::::::::::::::://
                  \\:::\^""--._.--""^/::://
                    \\::"\         /":://
                      \\::"\     /":://
                        \\::"\_/":://
                          \\::::://
                            \\_//
                              "
```

What's the problem :-)?

**Comment by nhaagen** [comment]
@klees dunno, force-pushed and apparently the tests are back

**Comment by klees** [comment]
Thx both =)

**Comment by mjansenDatabay** [comment]
You called me, and you didn't need me :(. I am feeling so useless^^.

**Comment by klees** [comment]
@mjansenDatabay I'm sorry. I thought that we would need your bright mind to solve this =)

**Comment by nhaagen** [comment]
picked https://github.com/ILIAS-eLearning/ILIAS/pull/6702

**Comment by klees** [comment]
Hi @Amstutz/@thibsy,

we started to incorporate this into our updated [cate](https://cate-tms.de) and could not find problems with the code until now. There are two PRs missing in here, but imho they contain marginal changes. So, from my perspective this is good to go for 8.

Would you be so kind and have a look into this now?

Kind regards!

File: src/UI/URLBuilder.php [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+
+        return $clone;
+    }
+
+    /**
+     * Change an acquired parameter's value if the supplied token is valid
+     */
+    public function withParameter(URLBuilderToken $token, $value): self
+    {
+        assert(ist_string($value) || is_array($value));
```
Comment (thibsy) [inline src/UI/URLBuilder.php]: I think this line has been added for backwards compatibility, right? Please don't use `assert()` though, and throw an `InvalidArgumentException` yourself instead. Also note the typo in `ist_string()`.

File: src/UI/Implementation/Render/AbstractComponentRenderer.php [hunk 2]
```diff
@@ -68,6 +72,7 @@ final public function __construct(
     public function registerResources(ResourceRegistry $registry): void
     {
         $registry->register('./src/UI/templates/js/Core/ui.js');
```
Comment (thibsy) [inline src/UI/Implementation/Render/AbstractComponentRenderer.php]: We don't need this file anymore, all functionality should be contained in `Core/dist/core.js` now. Please remove this line and delete the corresponding file as well.

**Review by thibsy — CHANGES_REQUESTED** [review]
Hi @nhaagen and @klees,

Thanks a lot for testing this with your (productive?) environment, and bundling all related changes together.

Since we have already reviewed the actual changes for ILIAS 9 and/or 10, we will only look at changes which might affect the back-porting to ILIAS 8 and PHP 7.4.

- [ ] Breaking changes: there are several breaking changes in there, what is your strategy to deal with them? To name some:
	- some input interface namespaces changed
	- the field selection view control interface and factory method changed
	- the view control interface was renamed and lost a method
	- the data table factory method changed
	- the data retrieval interface changed
	- the column interface changed
- [x] Inline comments: I have added some inline comments which need to be addressed. But maybe wait until we have discussed the first point at the next JF.

Kind regards,
@Amstutz and @thibsy

**Comment by matthiaskunkel** [comment]
Jour Fixe, 15 APR 2024: We still prefer to allow plugin developers to use UIDataTable in ILIAS8-plugins and therefore suggest to implement a solution for ILIAS 8 that allows input interfaces to stay in there previous namespaces. We accept breaking changes in interface to the data table and related components as they could not used in ILIAS 8 until now.

**Comment by nhaagen** [comment]
We added legacy interfaces for 

- ILIAS\UI\Component\Input\Field\Input
- ILIAS\UI\Component\Input\Field\FilterInput
- ILIAS\UI\Component\Input\ViewControl\ViewControl

and marked them as deprecated for 9.
That should strongly hint to necessary adjustments without breaking too much.

**Pull request merged**